### PR TITLE
remove deleted items from cache individually

### PR DESCRIPTION
**Pull request by howufiger7**
previously we were repopulating the cache by clearing it entirely first, which leads to inconsistencies in multi-threaded scenarios (e.g., thread A queries table while cache is being repopulated in thread B, so thread A appears to be missing data)

### CODE REVIEW

File: tests/test_omen.py [hunk 1]
```diff
@@ -931,3 +931,29 @@ def test_cache_sharing():
     db.delete("cars", id=45)
     cache.reload()
     assert not cache.select_one(id=45)
+
+
+def test_cache_sharing_threaded():
```
Comment (earonesty) [inline tests/test_omen.py:936]: probably should add a test where you remove one item from the db, and assert it's not in the cache

File: omen2/table.py [hunk 2]
```diff
@@ -126,13 +127,20 @@ def __select(self, where) -> Iterable[T]:
             if all(getattr(obj, k) == v for k, v in attr_where.items()):
                 yield obj
 
+        if not where:
```
[the remark this answers is not in the document]
Comment (howufiger7) [reply omen2/table.py:130]: so the previous impl cleared the cache when the where is empty, and that gets us into trouble in a threaded context (the new test i added is an example, don't even need 10 threads, it fails for 2). remove only the items that were deleted fixes that.

but you're right, this impl works more generally, even when there is a where filter specified, will update this.

File: tests/test_omen.py (hunk 1 above)
Comment (howufiger7) [reply tests/test_omen.py:936]: `test_cache_sharing` (above) does exactly that, but in a single thread. is that what you meant?

File: omen2/table.py [hunk 3]
```diff
@@ -106,13 +106,14 @@ def db_select(self, where):
         return self.db.select(self.table_name, None, where)
 
     def __select(self, where) -> Iterable[T]:
-        if not where:
```
Comment (earonesty) [inline omen2/table.py:109]: keep this

File: omen2/table.py [hunk 4]
```diff
@@ -106,13 +106,14 @@ def db_select(self, where):
...
-            self._cache.clear()
+        db_pks = set()
         db_where = {k: v for k, v in where.items() if k in self.field_names}
         attr_where = {k: v for k, v in where.items() if k not in self.field_names}
         for row in self.db_select(db_where):
             obj = self.row_type._from_db_not_new(row._asdict())
-            cached: "ObjBase" = self._cache.get(obj._to_pk_tuple())
+            pk = obj._to_pk_tuple()
+            cached: "ObjBase" = self._cache.get(pk)
+            db_pks.add(pk)
```
Comment (earonesty) [inline omen2/table.py:116]: if "where"

File: omen2/table.py [hunk 5]
```diff
@@ -126,13 +127,22 @@ def __select(self, where) -> Iterable[T]:
             if all(getattr(obj, k) == v for k, v in attr_where.items()):
                 yield obj
 
+        # remove cached items that are no longer in the db
+        remove_from_cache = set()
```
Comment (earonesty) [inline omen2/table.py:131]: if where:

File: tests/test_omen.py (hunk 1 above)
Comment (earonesty) [reply tests/test_omen.py:944]: yeah, i didn't see the case where there is no "where" clause set

**Pull request merged**